### ISSUE TITLE
apiclient.discovery build slow in certain servers

### ISSUE THREAD

**Issue by thassyo-kondado** [googleapis/google-api-python-client#957]
Thanks for stopping by to let us know something could be better!

**PLEASE READ**: If you have a support contract with Google, please create an issue in the [support console](https://cloud.google.com/support/) instead of filing on GitHub. This will ensure a timely response.

Please run down the following list and make sure you've tried the usual "quick fixes":

  - Search the issues already opened: https://github.com/googleapis/google-api-python-client/issues
  - Search StackOverflow: https://stackoverflow.com/questions/tagged/google-cloud-platform+python

If you are still having issues, please be sure to include as much information as possible:

#### Environment details

  - OS type and version: Multiple linux distros
  - Python version: `2.7`
  - pip version: `18.1`
  - `google-api-python-client` version: `1.9.3`

#### Steps to reproduce

  1. Run apiclient.discovery.build on different servers
  2. In some servers, the build process takes over 10 minutes just to return the client

#### Code example

```python
from apiclient.discovery import build
build(service, version, credentials=credentials)
```

#### Stack trace

This is a sudden error that started to occur a few weeks ago.
I have several servers where I run googleapis' build method, but I've started to notice that in some of them, without any change from my end, the build process started to take a really long time and even timeout in some cases.
I have started to use build_from_document instead of build, with a discovery document stored internally (even though the added code just to store the document seems too much for an error that is pretty inexplicable). However, there is another bottleneck in this process, which is the "credentials" part of the build function, which is taking a long time to complete as well and, since this is a google-maintained library which methods are used internally for the downstream code, I have not yet found a work around for it.
I have updated httplib2 to the latest version and googleapis as well, but the slowness and timeout continued for the affected servers - it is important to notice that the "slow" servers have the exact same configuration as the "fast" servers - using the same GCP project and client tokens.
This is happening regardless of the service (tested with bigquery, drive, analytics (v3 and v4), sheets, and google search console).
Besides using build_from_document, I also raised the socket timeout (socket.setdefaulttimeout) before calling the build method to prevent timeout exceptions, but the slowness is a major issue.

**Comment by busunkim96** [comment]
Is there anything else special about the environments you are working in? 

What kind of credentials are you using?

**Comment by thassyo-kondado** [comment]
Nothing special about the environments where there is a slow response.
Here is the credential flow we use:
```
from google.oauth2.credentials import Credentials as GCP_Credentials
credentials = GCP_Credentials(
            token,
            refresh_token=refresh_token,
            token_uri=token_uri,
            client_id=client_id,
            client_secret=client_secret
        )
build(service, version, credentials=credentials)
```
But again, the slow response is not only in the auth flow, but also to read the discovery docs

**Comment by busunkim96** [comment]
Hmm, the bottleneck is probably the HTTP requests then. Could you try enabling [logging](https://github.com/googleapis/google-api-python-client/blob/master/docs/logging.md#logging) and letting me know if you see anything special?

**Comment by thassyo-kondado** [comment]
Here is the response during **build**:

```
connect fail: (www.googleapis.com, 443)
connect fail: (www.googleapis.com, 443)
connect: (www.googleapis.com, 443)
send: 'GET /discovery/v1/apis/analyticsreporting/v4/rest HTTP/1.1\r\nHost: www.googleapis.com\r\naccept-encoding: gzip, deflate\r\nuser-agent: Python-httplib2/0.18.1 (gzip)\r\n\r\n'
reply: 'HTTP/1.1 200 OK\r\n'
header: Content-Type: application/json; charset=UTF-8
header: Vary: Origin
header: Vary: X-Origin
header: Vary: Referer
header: Content-Encoding: gzip
header: Date: Tue, 07 Jul 2020 13:35:00 GMT
header: Server: ESF
header: Cache-Control: private
header: X-XSS-Protection: 0
header: X-Frame-Options: SAMEORIGIN
header: X-Content-Type-Options: nosniff
header: Alt-Svc: h3-29=":443"; ma=2592000,h3-27=":443"; ma=2592000,h3-25=":443"; ma=2592000,h3-T050=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"
header: Transfer-Encoding: chunked
```

Then after a call to **service.reports().batchGet** (please notice that I omitted sensitive info with "XXXX...")
```
connect fail: (analyticsreporting.googleapis.com, 443)
connect: (analyticsreporting.googleapis.com, 443)
send: 'POST /v4/reports:batchGet?alt=json HTTP/1.1\r\nHost: analyticsreporting.googleapis.com\r\ncontent-length: 345\r\naccept-encoding: gzip, deflate\r\naccept: application/json\r\nuser-agent: (gzip)\r\ncontent-type: application/json\r\nauthorization: Bearer XXXXXXXXX\r\nx-goog-api-client: gdcl/1.9.3 gl-python/2.7.12\r\n\r\n{"reportRequests": [{"dateRanges": [{"startDate": "2020-06-03", "endDate": "2020-06-03"}], "dimensions": [{"name": "ga:date"}], "pageSize": 1000, "viewId": "XXXXXXXX", "metrics": [{"expression": "ga:goal6Completions"}, {"expression": "ga:goal19Completions"}, {"expression": "ga:goal20Completions"}], "pageToken": "0", "samplingLevel": "LARGE"}]}'
reply: 'HTTP/1.1 401 Unauthorized\r\n'
header: WWW-Authenticate: Bearer realm="https://accounts.google.com/", error="invalid_token"
header: Vary: Origin
header: Vary: X-Origin
header: Vary: Referer
header: Content-Type: application/json; charset=UTF-8
header: Content-Encoding: gzip
header: Date: Tue, 07 Jul 2020 13:37:10 GMT
header: Server: ESF
header: Cache-Control: private
header: X-XSS-Protection: 0
header: X-Frame-Options: SAMEORIGIN
header: X-Content-Type-Options: nosniff
header: Alt-Svc: h3-29=":443"; ma=2592000,h3-27=":443"; ma=2592000,h3-25=":443"; ma=2592000,h3-T050=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"
header: Transfer-Encoding: chunked
No handlers could be found for logger "google_auth_httplib2"
connect fail: (accounts.google.com, 443)
connect: (accounts.google.com, 443)
send: u'POST /o/oauth2/token HTTP/1.1\r\nHost: accounts.google.com\r\nContent-Length: 229\r\ncontent-type: application/x-www-form-urlencoded\r\naccept-encoding: gzip, deflate\r\nuser-agent: Python-httplib2/0.18.1 (gzip)\r\n\r\nclient_secret=XXXXXXX&grant_type=refresh_token&refresh_token=XXXXXXXX&client_id=XXXXXX'
reply: 'HTTP/1.1 200 OK\r\n'
header: Content-Type: application/json; charset=utf-8
header: Vary: Origin
header: Vary: X-Origin
header: Vary: Referer
header: Content-Encoding: gzip
header: Date: Tue, 07 Jul 2020 13:39:19 GMT
header: Server: scaffolding on HTTPServer2
header: Cache-Control: private
header: X-XSS-Protection: 0
header: X-Frame-Options: SAMEORIGIN
header: X-Content-Type-Options: nosniff
header: Alt-Svc: h3-29=":443"; ma=2592000,h3-27=":443"; ma=2592000,h3-25=":443"; ma=2592000,h3-T050=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"
header: Transfer-Encoding: chunked
send: 'POST /v4/reports:batchGet?alt=json HTTP/1.1\r\nHost: analyticsreporting.googleapis.com\r\ncontent-length: 345\r\naccept-encoding: gzip, deflate\r\naccept: application/json\r\nuser-agent: (gzip)\r\ncontent-type: application/json\r\nauthorization: Bearer XXXXXXXX\r\nx-goog-api-client: gdcl/1.9.3 gl-python/2.7.12\r\n\r\n{"reportRequests": [{"dateRanges": [{"startDate": "2020-06-03", "endDate": "2020-06-03"}], "dimensions": [{"name": "ga:date"}], "pageSize": 1000, "viewId": "XXXXXXX", "metrics": [{"expression": "ga:goal6Completions"}, {"expression": "ga:goal19Completions"}, {"expression": "ga:goal20Completions"}], "pageToken": "0", "samplingLevel": "LARGE"}]}'
reply: 'HTTP/1.1 200 OK\r\n'
header: Content-Type: application/json; charset=UTF-8
header: Vary: Origin
header: Vary: X-Origin
header: Vary: Referer
header: Content-Encoding: gzip
header: Date: Tue, 07 Jul 2020 13:39:19 GMT
header: Server: ESF
header: Cache-Control: private
header: X-XSS-Protection: 0
header: X-Frame-Options: SAMEORIGIN
header: X-Content-Type-Options: nosniff
header: Alt-Svc: h3-29=":443"; ma=2592000,h3-27=":443"; ma=2592000,h3-25=":443"; ma=2592000,h3-T050=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"
header: Transfer-Encoding: chunked
```

**Comment by Logic-gate** [comment]
I am facing the same exact issue and I've noticed that the connection fails 4 times every-time before the connection is made. At first I thought it was an issue with my code, but while testing with `httplib2`-- GET request to `googleapis.com` the same thing happens. I then tried `requests`....same thing. I tried `curl` and it worked. Mind you it's not a network issue, I tried the same script on various devices on the same connection and same versions of modules all worked just fine. Plus, `curl` worked.

Would appreciate any feedback....


```
credentialsCheck::tokenFound::/home/nemo/.config/harbour-cargo/token.pickle
credentialsCheck::goto::testFunction
serviceBuild::starting
serviceBuild::serviceBuilding
connect fail: (www.googleapis.com, 443)
connect fail: (www.googleapis.com, 443)
connect fail: (www.googleapis.com, 443)
connect fail: (www.googleapis.com, 443)
connect: (www.googleapis.com, 443)
...
...
...
reply: 'HTTP/1.1 200 OK\r\n'
header: Expires: Wed, 14 Oct 2020 14:53:40 GMT
header: Date: Wed, 14 Oct 2020 14:48:40 GMT
header: Cache-Control: public, max-age=300, must-revalidate, no-transform
header: Vary: Origin
header: Vary: X-Origin
header: Content-Type: application/json; charset=UTF-8
header: Content-Encoding: gzip
header: X-Content-Type-Options: nosniff
header: X-Frame-Options: SAMEORIGIN
header: Content-Security-Policy: frame-ancestors 'self'
header: X-XSS-Protection: 1; mode=block
header: Server: GSE
header: Alt-Svc: h3-Q050=":443"; ma=2592000,h3-29=":443"; ma=2592000,h3-27=":443"; ma=2592000,h3-T051=":443"; ma=2592000,h3-T050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"
header: Transfer-Encoding: chunked
...
...
...
connect fail: (www.googleapis.com, 443)
connect fail: (www.googleapis.com, 443)
connect fail: (www.googleapis.com, 443)
connect fail: (www.googleapis.com, 443)
connect: (www.googleapis.com, 443)
...
...
...
reply: 'HTTP/1.1 200 OK\r\n'
header: Expires: Wed, 14 Oct 2020 14:50:46 GMT
header: Date: Wed, 14 Oct 2020 14:50:46 GMT
header: Cache-Control: private, max-age=0, must-revalidate, no-transform
header: Vary: Origin
header: Vary: X-Origin
header: Content-Type: application/json; charset=UTF-8
header: Content-Encoding: gzip
header: X-Content-Type-Options: nosniff
header: X-Frame-Options: SAMEORIGIN
header: Content-Security-Policy: frame-ancestors 'self'
header: X-XSS-Protection: 1; mode=block
header: Server: GSE
header: Alt-Svc: h3-Q050=":443"; ma=2592000,h3-29=":443"; ma=2592000,h3-27=":443"; ma=2592000,h3-T051=":443"; ma=2592000,h3-T050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"
header: Transfer-Encoding: chunked
testFunction::resultsParse::items

```

**Comment by parthea** [comment]
Hi @thassyo-kondado,

We've just released version 2.0 of `google-api-python-client`. One of the key features is that `discovery.build()` will receive discovery documents from the client library itself rather than fetching them over the internet. Please try version 2.0 and let us know if you're still experiencing an issue. I'm going to close off this issue due to inactivity. If you're still having trouble, please re-open it with more information.
